### PR TITLE
[RFC 0008] Readonly recursive Nix

### PR DESCRIPTION
**Pull request by shlevy**
Rendered: https://github.com/shlevy/rfcs/blob/readonly-recursive-nix/rfcs/0008-readonly-recursive-nix.md

### CODE REVIEW

**Comment by shlevy** [comment]
@edolstra

File: rfcs/0008-readonly-recursive-nix.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+[FdDaemonStore][fd-daemon-store], and reading packets from the other.
+When connection requests are received, the build loop should create a
+new connection with the daemon, tell the daemon what it needs to know
+to set up a [StoreViewStore][store-view-store], and then pass the
+client socket through to the build.
+
+Multiple processes in the build may be trying to open connections in
+parallel, but as long as each only consumes one response it doesn't
+matter if the response is the exact one corresponding to the specific
+request made.
```
Comment (edolstra) [inline rfcs/0008-readonly-recursive-nix.md]: I don't understand what this means. It seems to say there is some sort of race with multiple connections but I don't see how.

**Comment by edolstra** [comment]
The idea seems mostly fine. A few comments:

* The main objection is that it significantly increases the attack surface from Nix builds, since they can now talk to a root process outside of the sandbox.

* Regarding daemon availability: A simpler approach would be to move all the daemon code from src/nix-daemon to libstore as a Daemon class that can be instantiated by DerivationGoal. So we just fire off a thread to handle incoming connections from the build. This also makes it easy to pass the StoreViewStore to the daemon (Daemon would just take a Store as a constructor argument).

* I don't see why FdDaemonStore is needed. Wouldn't the regular RemoteStore work? The only thing needed is a way to set the socket path (at least for non-sandbox builds where we can't create `/nix/var/nix/daemon-socket/socket`).

* I don't understand the name "StoreViewStore". Maybe RestrictedStore?

File: rfcs/0008-readonly-recursive-nix.md (hunk 1 above)
Comment (shlevy) [inline rfcs/0008-readonly-recursive-nix.md]: All processes in the build share the same socket FD they use to request a new connection, those requests could beinterleaved.

**Comment by shlevy** [comment]
@edolstra thanks for taking a look!

> The main objection is that it significantly increases the attack surface from Nix builds, since they can now talk to a root process outside of the sandbox.

True, I'll include that with the next draft (and the counterpoint that the daemon is already meant to be accessible to everyone in a normal setup)

> Regarding daemon availability: A simpler approach would be to move all the daemon code from src/nix-daemon to libstore as a Daemon class that can be instantiated by DerivationGoal. So we just fire off a thread to handle incoming connections from the build. This also makes it easy to pass the StoreViewStore to the daemon (Daemon would just take a Store as a constructor argument).

Sure, makes sense.

> I don't see why FdDaemonStore is needed. Wouldn't the regular RemoteStore work? The only thing needed is a way to set the socket path (at least for non-sandbox builds where we can't create /nix/var/nix/daemon-socket/socket).

Inheriting a file descriptor is just an alternative to specifying the socket path.I slightly prefer the former, but not strongly, I'll change this to do it that way instead if you prefer.

> I don't understand the name "StoreViewStore". Maybe RestrictedStore?

Yeah, that's btter. Though the more I think about it the more it seems like this is more a parameter of the daemon than an actual standalone store implementation anyway.

**Comment by joachifm** [comment]
> and the counterpoint that the daemon is already meant to be accessible to everyone in a normal setup

If I understand correctly, the concern is not necessarily other local users but that the proposal increases the feasibility of attacking the host by injecting malicious code into the build process.

**Comment by edolstra** [comment]
Right. Currently sandboxed builds don't have access to the daemon socket.

**Comment by shlevy** [comment]
Right, it's not a complete counterpoint, just that the daemon already should be pretty secure.

**Comment by shlevy** [comment]
We could put this behind an off-by-default option.

**Comment by Ericson2314** [comment]
I'd like to see some of the comparison with import-from-derivation fleshed out. For example, I'm missing why this is more expressive.

**Comment by shlevy** [comment]
@edolstra @Ericson2314 Updated in response to your comments.

**Comment by Ericson2314** [comment]
@slevy I'll respond in detail later but I think all those IFD drawbacks are very surmountable.

**Comment by shlevy** [comment]
@edolstra can we expect a response either way from you here?

**Comment by edolstra** [comment]
This looks good to me, thanks!

**Comment by shlevy** [comment]
@edolstra Awesome! :confetti_ball: 

@zimbatm So what's the next step from the perspective of the RFC process?

**Comment by Ericson2314** [comment]
@edolstra, would you also be OK in an RFC for improving import-from-derivation? They are in many cases competing ways to achieve the same thing, but implementation-wise there is no conflict.

**Comment by 0xABAB** [comment]
I think the RFC document (the one in the first message) is so vague that I have no idea what problems it intends to solve. Just based on that, I am already against.

**Comment by zimbatm** [comment]
@shlevy the RFC is very dry right now. @edolstra and you can understand it because you have a lot of shared context on the nix implementation.

To help the user's perspective, would you mind adding an example of nix code that would use the feature and then describe the various steps that the runtime would take when invoking the code?

**Comment by shlevy** [comment]
@zimbatm good call, will do when I get a chance

**Pull request closed without merging**